src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp

```diff
@@ -0,0 +1,299 @@
+    setDescription( "List of fluid phases" );
+
+  // 1) First option: specify PVT tables from one file per phase, read the files line by line, and populate the internal TableFunctions
+  registerWrapper( viewKeyStruct::tableFilesString(), &m_tableFiles ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setRestartFlags( RestartFlags::NO_WRITE ).
+    setDescription( "List of filenames with input PVT tables (one per phase)" );
+
+  // 2) Second option: specify TableFunction names for each phase,
+  registerWrapper( viewKeyStruct::densityTableNamesString(), &m_densityTableNames ).
+    setRTTypeName( rtTypes::CustomTypes::groupNameRefArray ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "List of density TableFuncion names from the Function block. \n"
+                    "The user must provide one TableFunction par phase, respecting the order provided in \"phaseNames\"." );
+
+  registerWrapper( viewKeyStruct::viscosityTableNamesString(), &m_viscosityTableNames ).
+    setRTTypeName( rtTypes::CustomTypes::groupNameRefArray ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "List of viscosity TableFuncion names from the Function block. \n"
+                    "The user must provide one TableFunction par phase, respecting the order provided in \"phaseNames\"." );
+
+  registerField( fields::twophasefluid::phaseDensity{}, &m_phaseDensity.value );
+  registerField( fields::twophasefluid::dPhaseDensity{}, &m_phaseDensity.derivs );
+  registerField( fields::twophasefluid::phaseDensity_n{}, &m_phaseDensity_n );
+
+  registerField( fields::twophasefluid::phaseViscosity{}, &m_phaseViscosity.value );
+  registerField( fields::twophasefluid::dPhaseViscosity{}, &m_phaseViscosity.derivs );
+}
+
+
+std::unique_ptr< ConstitutiveBase >
+TwoPhaseFluid::deliverClone( string const & name, Group * const parent ) const
+{
+  return ConstitutiveBase::deliverClone( name, parent );
+}
+
+
+void TwoPhaseFluid::resizeFields( localIndex const size, localIndex const numPts )
+{
+  // Assume sole dependency on pressure, i.e. one derivative
+  m_phaseDensity.value.resize( size, numPts, 2 );
+  m_phaseDensity.derivs.resize( size, numPts, 2, 1 );
+
+  m_phaseDensity_n.resize( size, numPts, 2 );
+
+  m_phaseViscosity.value.resize( size, numPts, 2 );
+  m_phaseViscosity.derivs.resize( size, numPts, 2, 1 );
+}
+
+
+void TwoPhaseFluid::allocateConstitutiveData( dataRepository::Group & parent,
+                                              localIndex const numConstitutivePointsPerParentIndex )
+{
+  ConstitutiveBase::allocateConstitutiveData( parent, numConstitutivePointsPerParentIndex );
+  resizeFields( parent.size(), numConstitutivePointsPerParentIndex );
+}
+
+
+void TwoPhaseFluid::postInputInitialization()
+{
+  ConstitutiveBase::postInputInitialization();
+
+  // Input relationships can be provided either as text files or TableFunctions.
+  m_tableFiles.empty() ? readInputDataFromTableFunctions() : readInputDataFromFileTableFunctions();
+
```

Comment:
Once the tables are created/read we need to check them for monotonicity. The density has to increase with pressure. Otherwise we will have a fluid with negative compressibility which might affect convergence down the line.

Reply:
Thanks for the heads up. I will add a `checkTableConsistency` as I have seen it implemented in other fluid models.